odoo/addons/test_testing_utilities/tests/test_freeze_time.py

```diff
@@ -0,0 +1,205 @@
+        real_after = fields.Datetime.now()
+
+        self.assertNotEqual(real_before, fake_time)
+        self.assertNotEqual(real_after, fake_time)
+        self.assertGreaterEqual(real_after, real_before)
+
+    def test_nested_freeze(self):
+        real_before = fields.Datetime.now()
+
+        with freeze_time('2018-12-03 09:38:10'):
+            fake_before = fields.Datetime.now()
+
+            with freeze_time('2018-12-03 09:40:00'):
+                fake_fake = fields.Datetime.now()
+
+            fake_after = fields.Datetime.now()
+
+        real_after = fields.Datetime.now()
+
+        self.assertEqual(fake_before, fake_after)
+        self.assertNotEqual(fake_before, fake_fake)
+        self.assertNotEqual(fake_fake, real_before)
+        self.assertNotEqual(fake_fake, real_after)
+        self.assertGreaterEqual(real_after, real_before)
+
+    def test_same_context_freeze(self):
+        f1 = freeze_time('2012-03-07')
+        f2 = freeze_time('2015-01-01')
+
+        with f1:
+            t1 = datetime.now()
+            with f2:
+                t2 = datetime.now()
+            t3 = datetime.now()
+
+        self.assertEqual(t1, t3)
+        self.assertNotEqual(t1, t2)
+
+    def test_datetime_to_date_freeze(self):
+        with freeze_time('2018-05-05 09:12:55'):
+            dt = datetime.now()
+            d = dt.date()
+
+        manual_d = date(2018, 5, 5)
+        self.assertEqual(d, manual_d)
+
+    def test_datetime_addition(self):
+        with freeze_time('2018-05-05 09:12:55'):
+            dt = datetime.now()
+            dt2 = dt + timedelta(hours=2)
+
+        manual_dt = datetime(2018, 5, 5, 11, 12, 55)
+        self.assertEqual(dt2, manual_dt)
+
+    def test_datetime_subtraction(self):
+        with freeze_time('2018-05-05 09:12:55'):
+            dt = datetime.now()
+            dt3 = dt - timedelta(hours=2)
+            dt4 = dt - dt
+
+        manual_dt = datetime(2018, 5, 5, 7, 12, 55)
+        self.assertEqual(dt3, manual_dt)
+        self.assertEqual(dt4, timedelta(0))
+
+    def test_astimezone_freeze(self):
+        gmt = timezone('GMT')
+        gmt2 = timezone('Etc/GMT+2')
+
+        with freeze_time('2018-05-05 09:12:53'):
+            dt = datetime.now(gmt2)
+            dt2 = dt.astimezone(gmt)
+
+        manual_dt_gmt = datetime(2018, 5, 5, 9, 12, 53, tzinfo=gmt)
+        self.assertEqual(dt2.tzinfo, gmt)
+        self.assertEqual(dt2, manual_dt_gmt)
+
+    def test_date_addition(self):
+        with freeze_time('2018-05-05'):
+            d = date.today()
+            d2 = d + timedelta(days=1)
+
+        manual_d = date(2018, 5, 6)
+        self.assertEqual(d2, manual_d)
+
+    def test_date_subtraction(self):
+        with freeze_time('0001-01-02'):
+            d = date.today()
+            d2 = d - timedelta(days=1)
+            d3 = d - d
+
+        manual_d = date.min
+        self.assertEqual(d2, manual_d)
+        self.assertEqual(d3, timedelta(0))
```

Comment:
I don't see the purpose of using `date.min` here.
<nitpicking>BTW, is `date.min` a `PatchedDate`?</nitpicking>

Reply:
`date.min == date(1, 1, 1)` which is what `date(1, 1, 2) - timedelta(days=1)` equals to, `date.min` is a `datetime.date` object as it's declared outside the context of `freeze_time`, but `d2` is a `PatchedDate` object.

The goal of this test is to test the possible outcomes of `__sub__` in a frozen time context